### DISCUSSION TITLE
Add C++ feature

### DISCUSSION THREAD

**Discussion by mering** [devcontainers-community/features#36] [Ideas]
This should basically contain https://github.com/devcontainers/images/blob/main/src/cpp/.devcontainer/Dockerfile.

Maybe align with [CMake](https://github.com/devcontainers-community/features-cmake) and [LLVM](https://github.com/devcontainers-community/features-llvm) features.

Fixes devcontainers/images#672

**Comment by jcbhmr** [comment]
Sounds like a good idea to use the cpp base image as a guide! 👍 Similar to the Python image https://github.com/devcontainers/images/blob/main/src/python/.devcontainer/devcontainer.json which has many devcontainer `"features": { ... }` in it to build the single `python` image, I think it might be good to split that C++ image into multiple features instead of one big C++ feature? 🤔 Then again, I'm not a big C++ user, so my opinion doesn't mean much. 🤷‍♀️

maybe
- llvm feature https://llvm.org/
- GCC feature https://gcc.gnu.org/
- cmake feature https://cmake.org/
- vcpkg feature https://vcpkg.io/en/
- conan feature https://conan.io/

(those are the big c++ tools that i used in university)

But I take it you know what you're doing (you did propose this after all 🤣) so **Go for it! 🚀** See how it pans out!

<sup>p.s. I've invited you to the org. you should be able to create said features 😉</sup>

**Reply by jcbhmr** [reply]
oh and btw i tried to explain some of the decisions behind like "why polyrepo vs monorepo" and such here: https://github.com/devcontainers-community/.github/blob/main/ARCHITECTURE.md and here https://github.com/devcontainers-community/features/discussions/34 that you might be interested in

**Reply by mering** [reply]
For many use cases the local `apt install` method is sufficient and is much faster. So features should have a `version: "apt"` option as well.

**Reply by jcbhmr** [reply]
> So features should have a version: "apt" option

👍 good idea! there appears to be an existing convention of `system` or `os-provided`: the python feature https://github.com/devcontainers/features/blob/main/src/python/install.sh#L10 uses `os-provided` or `system` to delegate to apt

**Reply by msclock** [reply]
That's right. So some downstreams features can be delegated by the upstream ones and the feild 'installsAfter' could be added to devcontainer feature configureation. But that field does not constrain upstream features to be installed automatically correctly, and perhaps requires some priority configuration or error checks.

**Comment by jcbhmr** [comment]
after some searching https://github.com/search?q=vcpkg+feature&type=registrypackages theres already an existing vcpkg feature by @msclock at https://github.com/msclock/features!

there's also quite a few gcc features https://github.com/search?q=gcc+feature&type=registrypackages which is cool

no conan.io features tho https://github.com/search?q=conan+feature&type=registrypackages so maybe its not as popular a tool as i though 🤷‍♀️

**Reply by mering** [reply]
btw: you should use `git filter-repo` to only extract the history of a specific sub-folder as explained in https://docs.github.com/en/get-started/using-git/splitting-a-subfolder-out-into-a-new-repository.
Please also invite the original author to that repo.

**Reply by msclock** [reply]
@jcbhmr hallo,I'm ok but be careful when i implementing vcpkg, consider the following things. 
- consider using only the gcc/g++ compiler that comes with the system. Of course, it is more flexible if you can configure the installation compiler to feature. But considering that you only need to choose a different linux distribution, you can also switch to a different compiler. So gcc feature wasn't on my implementation list in the first place. 
- when cmake does not exist, use vcpkg to automatically download the appropriate version. So cmake feature is not on my list either.

And,I would also like to join the community to maintain it. 😃

**Reply by mering** [reply]
@msclock I invited you to https://github.com/devcontainers-community/features-vcpkg. Great to see you onboard maintaining it going forward!

**Reply by jcbhmr** [reply]
@mering oh cool! didnt know that filter-repo was a thing!  from some googling it looks like there's multiple ways to do it? https://making.close.com/posts/splitting-sub-folders-out-into-new-git-repository

**Reply by mering** [reply]
Yes. As you can read in the article, even git documentation suggests to use `git filter-repo` instead of `git filter-branch`. `git subtree` is a weird thing imho especially as it seems to link the old and the new subtrees somehow together which is not what we want in this case. So I suggest `git filter-repo`.
